File: Medium/79/79.ts
export {};

const DIRECTIONS = [
  [-1, 0],
  [0, 1],
  [1, 0],
  [0, -1],
];

function dfs(
  board: string[][],
  word: string,
  m: number,
  n: number,
  x: number,
  y: number,
  idx: number,
): boolean {
  if (idx == word.length) return true;
  if (x < 0 || x >= m || y < 0 || y >= n) return false;
  if (board[x][y] !== word[idx]) return false;
  const temp = board[x][y];
  board[x][y] = "\0";
  for (const v of DIRECTIONS)
    if (dfs(board, word, m, n, x + v[0], y + v[1], idx + 1)) {
      board[x][y] = temp;
      return true;
    }
  board[x][y] = temp;

  return false;
}
function exist(board: string[][], word: string): boolean {
  const m = board.length;
  const n = board[0].length;
  for (let i = 0; i < m; i++)
    for (let j = 0; j < n; j++)
      if (dfs(board, word, m, n, i, j, 0)) return true;
  return false;
}

console.log("It's Showtime Folks!");
let board = [
  ["A", "B", "C", "E"],
  ["S", "F", "C", "S"],
  ["A", "D", "E", "E"],
];
let word = "ABCCED";
console.log(`Result: ${exist(board, word)}`);
board = [
  ["A", "B", "C", "E"],
  ["S", "F", "C", "S"],
  ["A", "D", "E", "E"],
];
word = "SEE";
console.log(`Result: ${exist(board, word)}`);
board = [
  ["A", "B", "C", "E"],
  ["S", "F", "C", "S"],
  ["A", "D", "E", "E"],
];
word = "ABCB";
console.log(`Result: ${exist(board, word)}`);

File: Medium/79/79.py
DIRECTIONS = [[-1, 0], [0, 1], [1, 0], [0, -1]]


def dfs(
    board: list[list[str]], word: str, m: int, n: int, x: int, y: int, idx: int
) -> bool:
    if idx == len(word):
        return True
    if x < 0 or x >= m or y < 0 or y >= n:
        return False
    if board[x][y] != word[idx]:
        return False
    temp = board[x][y]
    board[x][y] = "\0"
    for v in DIRECTIONS:
        if dfs(board, word, m, n, x + v[0], y + v[1], idx + 1):
            board[x][y] = temp
            return True
    board[x][y] = temp
    return False


class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        for i in range(m):
            for j in range(n):
                if dfs(board, word, m, n, i, j, 0):
                    return True
        return False


print("It's Showtime Folks!")
board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
word = "ABCCED"
print(f"Result: {Solution().exist(board, word)}")
board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
word = "SEE"
print(f"Result: {Solution().exist(board, word)}")
board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
word = "ABCB"
print(f"Result: {Solution().exist(board, word)}")
